libgadget/utils/system.c

```diff
@@ -453,6 +528,50 @@ get_physmem_bytes()
     return 64 * 1024 * 1024;
 }
 
+int
+_MPIU_Barrier(const char * fn, const int line, MPI_Comm comm)
+{
+    int ThisTask, NTask;
+    MPI_Comm_size(comm, &NTask);
+    MPI_Comm_rank(comm, &ThisTask);
+    int * recvbuf = malloc(sizeof(int) * NTask);
+    MPI_Request request;
+    int flag = 0;
+    int tag = 0;
+    int i;
+    for(i = 0; i < strlen(fn); i ++) {
+        tag += (int)fn[i] * 8;
+    }
+    tag += line;
+
+    MPI_Igather(&tag, 1, MPI_INT, recvbuf, 1, MPI_INT, 0, comm, &request);
+// #ifdef DEBUG TODO if this is a performance drag, change it to debug only
+    i = 0;
+    while(flag) {
+        MPI_Test(&request, &flag, MPI_STATUS_IGNORE);
+        usleep(50);
```

Comment:
Can we make the performance hit less by gradating the usleep? eg usleep(MIN(80, 2 * i))? Or similar?